### ISSUE TITLE
IPython doesn't clear its own flags from `sys.argv`

### ISSUE THREAD

**Issue by oconnor663** [ipython/ipython#9133]
If I run something like `ipython --no-confirm-exit`, the flag there is still visible in `sys.argv`:

```
$ ipython --no-confirm-exit
Python 3.5.1 (default, Dec  7 2015, 12:58:09) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
Jack's startup.py...

In [1]: sys.argv
Out[1]: ['/usr/bin/ipython', '--no-confirm-exit']
```

This confused the Kivy library, for example: https://github.com/kivy/kivy/issues/3918

**Comment by takluyver** [comment]
I don't think there's anything wrong here. Programmes are not expected to remove arguments they recognise from `sys.argv` - argparse, the standard Python library for handling command line options, does not do this either. Libraries are not expected to jump in and start processing `sys.argv` on import. Kivy is decidedly unusual in doing this.

You can work around this by running this before importing Kivy:

```python
sys.argv[1:] = []
```

**Comment by oconnor663** [comment]
Agreed that the IPython behavior is kind of non-standard. For comparison, though, the regular `python` interpreter does clear its own flags out of the args:

```
$ python -W ignore
Python 3.5.1 (default, Dec  7 2015, 12:58:09) 
[GCC 5.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.argv
['']
```

...sometimes. The `-c` flag seems to be a special flower in this respect.

**Comment by kived** [comment]
Kivy is an application framework, not just a library. It does have its own command line arguments, and we can't possibly know what arguments are supposed to be for IPython. This would apply equally to any program which parses arguments - it will see IPython's args and try to parse them.

Also, note that IPython does indeed strip those args if you run a script from the command line like `ipython --no-confirm-exit script.py` or `ipython script.py --no-confirm-exit`, and the only arguments which remain in `sys.argv` are ones which appear after `--`, i.e. `ipython --no-confirm-exit script.py -- --hello`. So this behavior seems inconsistent.

**Comment by Carreau** [comment]
It's impossible to know all flag accepted by IPython as any `--foo.Bar` argument might be valid.

I dont think `ipython --no-confirm-exit script.py` vs `ipython script.py --no-confirm-exit` is inconstant.
It's how argparse and subarguments works (as well as other non-python libraries) , and it can perfectly be read as `(interpreter with flag) (program) (arguments)`,

**Comment by oconnor663** [comment]
@Carreau, I believe what @kived was pointing out was that the running-a-script behavior is different from the interactive-interpreter behavior, regardless of how the arguments are ordered. For example, if I write this program:

```python
import sys
print(sys.argv)
```

and run it with `ipython test.py --no-confirm-exit`, it prints out something like `['/tmp/test.py']`. Which is to say, it *does* strip the `--no-confirm-exit` argument from the `sys.argv` list, *unlike* the behavior we see in the interactive IPython interpreter.

I had a typo [above](https://github.com/ipython/ipython/issues/9133#issuecomment-172109773) where I meant to agree with @takluyver, that it's surprising to me that Kivy reads the contents of `sys.argv` on import, rather than waiting for me to make some explicit function call into the library, where I might be able to supply my own arguments.

**Comment by kived** [comment]
@oconnor663 Exactly. That's how I tested as well. The only way to get arguments to a script run via IPython command line is by using `--` to separate IPython arguments from script arguments.

I agree that parsing `sys.argv` on import isn't great, but that's also irrelevant. The point is that **if** we parse `sys.argv`, we will have this issue. When we parse isn't important.

Remember that Kivy is an application framework, not a library. In fact, we have almost the same behavior as pygtk in this regard.
